[Swift in WebKit] Work towards proper WebGPU modularization (part 1)

rr-codes opened the pull request:
#### ccabb81a323018bdf1f9985be5659da5360065a2
<pre>
[Swift in WebKit] Work towards proper WebGPU modularization (part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=302588">https://bugs.webkit.org/show_bug.cgi?id=302588</a>
<a href="https://rdar.apple.com/164821471">rdar://164821471</a>

Reviewed by BJ Burg and Mike Wyrzykowski.

- Add missing headers to the Xcode project
- Remove unnessecary Logging files and functions
- Properly mark the internal module umbrella header as an umbrella header
- The internal module umbrella header should only contain includes of project headers; remove all others
- Move all declarations from the umbrella header to a regular header file instead
- Remove unnecessary declarations, including `WTFRangeSizeT`, `stdDynamicExtent`, `WGPU_COPY_STRIDE_UNDEFINED_`, and `MTLBlitOptionNone_`
- Change the namespace name to `CxxBridging` since the namespace name `WebGPU_Internal` doesn&apos;t really make sense or reflect its purpose
- Remove `checkedDifferenceSizeT` since it wasn&apos;t doing anything in actuality
- Remove `clampDouble` in favor of a native Swift function
- Fix some inconsistent formatting

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.mm:
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.getMappedRange(_:size:)):
(computeRangeSize(_:offset:)): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.finish(_:)):
(WebGPU.options):
(WebGPU.errorString(_:)):
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingCopyTextureToBuffer(_:destination:copySize:)):
(WebGPU.errorValidatingImageCopyBuffer(_:)):
(WebGPU.errorValidatingCopyBufferToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingRenderPassDescriptor(_:)):
(WebGPU.errorValidatingTimestampWrites(_:)):
(WebGPU.isMultisampleTexture(_:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.hasValidDimensions(_:width:height:depth:)):
(WebGPU.copyBufferToBuffer(_:sourceOffset:destination:destinationOffset:size:)):
(WebGPU.copyTextureToBuffer(_:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
(WebGPU.clearBuffer(_:offset:size:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
(WebGPU.beginComputePass(_:)):
(CommandEncoder_beginRenderPass_thunk(_:descriptor:)):
(CommandEncoder_beginComputePass_thunk(_:descriptor:)):
(CommandEncoder_runClearEncoder_thunk(_:attachmentsToClear:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)): Deleted.
(CommandEncoder_finish_thunk(_:descriptor:)):
(WebGPU.clearTextureIfNeeded(_:_:_:_:_:)): Deleted.
(WebGPU.runClearEncoder(_:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)): Deleted.
(WebGPU.resolveQuerySet(_:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
(WebGPU.validateResolveQuerySet(_:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
* Source/WebGPU/WebGPU/CxxBridging.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
(roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong):
(roundUpToMultipleOfNonPowerOfTwoUInt32UInt32):
(CxxBridging::isValidToUseWithTextureViewCommandEncoder):
(CxxBridging::isValidToUseWithQuerySetCommandEncoder):
(CxxBridging::isValidToUseWithBufferCommandEncoder):
(CxxBridging::isValidToUseWithTextureCommandEncoder):
(CxxBridging::isValidToUseWith):
(CxxBridging::areBuffersEqual):
(CxxBridging::convertWTFStringToNSString):
(CxxBridging::commandBufferThreadSafeWeakPtr):
* Source/WebGPU/WebGPU/Internal/Logging.cpp: Removed.
* Source/WebGPU/WebGPU/Internal/Logging.h: Removed.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::logString): Deleted.
(WebGPU_Internal::roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong): Deleted.
(WebGPU_Internal::roundUpToMultipleOfNonPowerOfTwoUInt32UInt32): Deleted.
(WebGPU_Internal::checkedDifferenceSizeT): Deleted.
(WebGPU_Internal::isValidToUseWithTextureViewCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithQuerySetCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithBufferCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithTextureCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWith): Deleted.
(WebGPU_Internal::clampDouble): Deleted.
(WebGPU_Internal::areBuffersEqual): Deleted.
(WebGPU_Internal::convertWTFStringToNSString): Deleted.
(WebGPU_Internal::commandBufferThreadSafeWeakPtr): Deleted.
* Source/WebGPU/WebGPU/Internal/module.modulemap:
* Source/WebGPU/WebGPU/Queue.mm:
* Source/WebGPU/WebGPU/Queue.swift:
* Source/WebGPU/WebGPU/StdLibExtras.swift:
(Comparable.clamped(to:)):

Canonical link: <a href="https://commits.webkit.org/303206@main">https://commits.webkit.org/303206@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b9126fc64a5a9fb35ef6f7f3c8b22d3107a356b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131674 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42641 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139183 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/41cd12e5-1b5a-4fcb-b5e0-9e998aca4594) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3928 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/06351af4-2f36-4521-8866-ee258641159b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134620 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/117948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81341 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2605044d-d21e-4574-9ecc-bf3f64cb30f7) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82374 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141829 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/3834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36640 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109025 "Failed to checkout and rebase branch from PR 54003") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3915 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109154 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27645 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/2939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114225 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20475 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/3888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67337 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/3849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1430,79 +1539,57 @@ extension WebGPU.CommandEncoder {
-                        return
-                    }
-                    blitCommandEncoder.copy(
+                var destinationOffset = UInt(destination.layout.offset)
```
idk why git made a mess of this part of the diff but all that’s different is the indentation...

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -309,10 +349,41 @@ extension WebGPU.CommandEncoder {
+            )
+
+        if !options.isEmpty {
+            // FIXME:  maybe it needs to be another value that is not sizeof.
```
I don't understand the above comment, 8 bits is always sizeof(Float) * 8 bits right? Or should we just divide by 4?

rr-codes replied:
I didn’t write that comment so I also have no idea 😛

mwyrzykowski replied:
Specifically the only place options is not empty is:
```
        var options: MTLBlitOption = []
        if mtlTexture!.pixelFormat == .depth32Float_stencil8 {
            options = .depthFromDepthStencil
        }
```
where there is a 32-bit depth component and an 8-bit stencil component

rr-codes replied:
(the git diff for this section is horrible and confusing for some reason)

rr-codes replied:
(this logic corresponds to the “deleted” code on line 298 of the old side of the diff; no logic has _actually_ changed)

mwyrzykowski replied:
Ah sorry got confused by the diff 👍 - my bad

rr-codes replied:
honesty its so confusing ill double check to make sure I didn’t do something silly

rr-codes replied:
ok ya it is correct, it is just mind boggling-y weirdly diff’d

mwyrzykowski commented on Source/WebGPU/WebGPU/Buffer.swift:
```diff
@@ -46,32 +46,22 @@ public func Buffer_getMappedRange_thunk(_ buffer: WebGPU.Buffer, offset: Int, si
 
         var rangeSize = size
         if size == WGPU_WHOLE_MAP_SIZE {
-            rangeSize = computeRangeSize(size: Int(currentSize()), offset: offset)
```
This is a behavior change when invalid offsets are passed on valid buffers and possibly in other scenarios, specifically offsets greater than the buffer size.

Previously, `validateGetMappedRange` would fail due to (in Buffer.mm):

```
    if (offset < m_mappingRange.beginOffset)
        return false;
```

now it will terminate the GPU process. I think this will cause CTS failures. EWS runs the objective-C++ version of WebGPU due to the Swift compiler version check in HardwareCapabilities.mm

rr-codes replied:
when would

```
if result.hasOverflowed() 
```

be true though? since both parameters are of type `Int`, and `Int` - `Int` can never overflow right?

rr-codes replied:
oh, is that function basically just a fancy

```
guard offset < size
```

check because of the implicit casting to size_t

mwyrzykowski replied:
oh you are correct I think, `Int` - `Int` can't overflow but it no longer clamps to 0 / now `rangeSize` can be negative. But then `rangeSize` is directly passed to:
```
Buffer::validateGetMappedRange(size_t offset, size_t rangeSize)
```
which expects an unsigned size_t type, so after casting `rangeSize` can now be greater than the `bufferSize` which is unexpected.

mwyrzykowski replied:
It's unfortunate we can directly pass an `Int` in Swift to a C++ member function expecting a `size_t` without a warning / error :(

rr-codes replied:
ya; is C++ `int` convertible to `size_t`? If it is I wonder if there’s some clang warning or something that would prevent that, and might also apply to Swift-Cxx interop?

rr-codes replied:
oh looks like there’s CLANG_WARN_IMPLICIT_SIGN_CONVERSION , which we don’t set anywhere

rr-codes replied:
though, confusingly, there is also GCC_WARN_SIGN_COMPARE ; not sure what the difference is

rr-codes replied:
enabling CLANG_WARN_IMPLICIT_SIGN_CONVERSION in WTF, WebGPU, and the other projects reveals quiteeeeeee a few issues 😅

mwyrzykowski replied:
Ok outside the scope of this PR then 👍 with the exception of ensuring the range check is correct

mwyrzykowski reviewed (commented):
Can you please manually run-webkit tests on the following two paths:
```
./LayoutTests/http/tests/webgpu
./LayoutTests/fast/webgpu
```
and let me know if there is anything other than all pass?

Due to the Swift compiler version check in HardwareCapabilities.mm, EWS runs the Objective-C++ version

rr-codes committed:
[Swift in WebKit] Work towards proper WebGPU modularization (part 1)
https://bugs.webkit.org/show_bug.cgi?id=302588
rdar://164821471

Reviewed by BJ Burg and Mike Wyrzykowski.

- Add missing headers to the Xcode project
- Remove unnessecary Logging files and functions
- Properly mark the internal module umbrella header as an umbrella header
- The internal module umbrella header should only contain includes of project headers; remove all others
- Move all declarations from the umbrella header to a regular header file instead
- Remove unnecessary declarations, including `WTFRangeSizeT`, `stdDynamicExtent`, `WGPU_COPY_STRIDE_UNDEFINED_`, and `MTLBlitOptionNone_`
- Change the namespace name to `CxxBridging` since the namespace name `WebGPU_Internal` doesn't really make sense or reflect its purpose
- Remove `checkedDifferenceSizeT` since it wasn't doing anything in actuality
- Remove `clampDouble` in favor of a native Swift function
- Fix some inconsistent formatting

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.mm:
* Source/WebGPU/WebGPU/Buffer.swift:
(WebGPU.getMappedRange(_:size:)):
(computeRangeSize(_:offset:)): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.finish(_:)):
(WebGPU.options):
(WebGPU.errorString(_:)):
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingCopyTextureToBuffer(_:destination:copySize:)):
(WebGPU.errorValidatingImageCopyBuffer(_:)):
(WebGPU.errorValidatingCopyBufferToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingRenderPassDescriptor(_:)):
(WebGPU.errorValidatingTimestampWrites(_:)):
(WebGPU.isMultisampleTexture(_:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.hasValidDimensions(_:width:height:depth:)):
(WebGPU.copyBufferToBuffer(_:sourceOffset:destination:destinationOffset:size:)):
(WebGPU.copyTextureToBuffer(_:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
(WebGPU.clearBuffer(_:offset:size:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
(WebGPU.beginComputePass(_:)):
(CommandEncoder_beginRenderPass_thunk(_:descriptor:)):
(CommandEncoder_beginComputePass_thunk(_:descriptor:)):
(CommandEncoder_runClearEncoder_thunk(_:attachmentsToClear:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)): Deleted.
(CommandEncoder_finish_thunk(_:descriptor:)):
(WebGPU.clearTextureIfNeeded(_:_:_:_:_:)): Deleted.
(WebGPU.runClearEncoder(_:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)): Deleted.
(WebGPU.resolveQuerySet(_:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
(WebGPU.validateResolveQuerySet(_:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
* Source/WebGPU/WebGPU/CxxBridging.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
(roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong):
(roundUpToMultipleOfNonPowerOfTwoUInt32UInt32):
(CxxBridging::isValidToUseWithTextureViewCommandEncoder):
(CxxBridging::isValidToUseWithQuerySetCommandEncoder):
(CxxBridging::isValidToUseWithBufferCommandEncoder):
(CxxBridging::isValidToUseWithTextureCommandEncoder):
(CxxBridging::isValidToUseWith):
(CxxBridging::areBuffersEqual):
(CxxBridging::convertWTFStringToNSString):
(CxxBridging::commandBufferThreadSafeWeakPtr):
* Source/WebGPU/WebGPU/Internal/Logging.cpp: Removed.
* Source/WebGPU/WebGPU/Internal/Logging.h: Removed.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::logString): Deleted.
(WebGPU_Internal::roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong): Deleted.
(WebGPU_Internal::roundUpToMultipleOfNonPowerOfTwoUInt32UInt32): Deleted.
(WebGPU_Internal::checkedDifferenceSizeT): Deleted.
(WebGPU_Internal::isValidToUseWithTextureViewCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithQuerySetCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithBufferCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWithTextureCommandEncoder): Deleted.
(WebGPU_Internal::isValidToUseWith): Deleted.
(WebGPU_Internal::clampDouble): Deleted.
(WebGPU_Internal::areBuffersEqual): Deleted.
(WebGPU_Internal::convertWTFStringToNSString): Deleted.
(WebGPU_Internal::commandBufferThreadSafeWeakPtr): Deleted.
* Source/WebGPU/WebGPU/Internal/module.modulemap:
* Source/WebGPU/WebGPU/Queue.mm:
* Source/WebGPU/WebGPU/Queue.swift:
* Source/WebGPU/WebGPU/StdLibExtras.swift:
(Comparable.clamped(to:)):

Canonical link: https://commits.webkit.org/303206@main
